### PR TITLE
Implement generic batch normalization layer and add most (if not all) tests.

### LINKED ISSUE #89116: Implement generic batch normalization layer.
### 🚀 The feature, motivation and pitch

Currently, the SyncBatchNorm module in PyTorch only has a `convert_sync_batchnorm` method to convert other Batch Normalization layers to SyncBatchNorm. However, it does not implement any method to revert the layers to their original classes.

However, implementing this would require either saving the original class type inside the SyncBatchNorm or implementing 1d, 2d, and 3d implementations of SyncBatchNorm. It would be preferable to implement instead a generic BatchNorm layer that can handle arbitrary dimensions.

Therefore, I would like to propose a generic `BatchNorm` layer that can handle arbitrary dimensions and can convert any related BatchNorm layer to itself, similar to `convert_sync_batchnorm`. This would allow removing the SyncBatchNorm layer for non-distributed inference and can convert any previous models to the new BatchNorm layer. Since the current functional `batch_norm` function is dimension neutral, the implementation will not be complicated.

### Alternatives

Possibly save the original dimensions in the SyncBatchNorm layer. However, this would not work if the original model is unknown.

### Additional context

SyncBatchNorm causes issues on single-process CPU inference and is challenging to convert to ONNX, etc. Having a generic BatchNorm would be the minimally invasive approach. Other issues have also proposed this solution before.

https://github.com/pytorch/pytorch/issues/41081
https://github.com/pytorch/pytorch/issues/82464

### PR DESCRIPTION
**Pull request by veritas9872**
Fixes #89116 by implementing a generic batch normalization layer and lazy version.
Quantized batch normalization has not been implemented.
All tests for lazy batch normalization have been implemented but not for batchnorm because of tight pairing with dimension number in many tests.
Open to feedback concerning tests.

### CODE REVIEW

**Comment by veritas9872** [comment]
As a first-time contributor, I am very open to any feedback. Please correct me if any mistakes have been made or additional work is required.

**Comment by veritas9872** [comment]
Might be related https://github.com/pytorch/pytorch/pull/89706

**Comment by veritas9872** [comment]
On second thoughts, it may be better to simply add a new class method `convert_to_batchnorm` that takes any child class of `_BatchNorm` as an input and convert all other child classes of `_BatchNorm` into the provided class. 

This way, there would be no additional `BatchNorm` class.

This would be feasible because, in the vast majority of cases, models have batch normalization layers with the same dimension.

**Comment by veritas9872** [comment]
This could be used to convert all SyncBatchNorm layers to BatchNorm3d layers for example.
